R/clean_taxon_names.R

```diff
@@ -0,0 +1,188 @@
+        )
+    )
+
+    # check that the data input is a data.frame or a tibble
+    assert_that(
+        is.data.frame(data) | dplyr::is.tbl(data),
+        msg = paste(data, "is not a data.frame or tibble object")
+    )
+
+    # check that the target_taxon column is in the data object
+    assert_that(
+        assertthat::is.string(target_taxon) & (target_taxon %in% names(data)),
+        msg = paste(target_taxon, "is not a column in the supplied data object")
+    )
+
+    # check that the name column has a length of more than zero and that it is
+    # a character variable
+    data_target_taxon <- data[[target_taxon]]
+    assert_that(
+        is.character(data_target_taxon) & (length(data_target_taxon) > 0),
+        msg = paste(
+            data_target_taxon,
+            "is not a character variable with length greater than zero"
+        )
+    )
+
+    # check that the life-stage column is a character vector without NAs
+    # TODO: this doesn't seem correct, there's NA in the list
+    data_life_stage <- data[[life_stage]]
+    assert_that(
+        (is.character(data_life_stage) & all(data_life_stage %in% c(
+            NA, "none", "larva", "pupa", "nymph",
+            "adult", "nauplius", "copepodite", "tadpole"
+        ))),
+        msg = paste(
+            data_life_stage,
+            "one or more entries do not have appropriate life-stage classes: see documentation" # TODO: not yet in the docs
+        )
+    )
+
+    # add a targ_no column to the data input
+    data[["row_id"]] <- 1:nrow(data)
+
+    # clean the names for typos etc. using the bdc_clean_names function
+    clean.names <- bdc::bdc_clean_names(
+        sci_names = data[[target_taxon]],
+        save_outputs = FALSE
+    )
+
+    # write some code to remove the output file
+    unlink("Output", recursive = TRUE)
+
+    # add the clean names to the data.frame
+    clean.col <- paste("clean_", target_taxon, sep = "")
+    data[[clean.col]] <- clean.names$names_clean
+
+    # check that all the non-missing names were cleaned
+    assert_that(
+        length(data_target_taxon) == length(data[[clean.col]]),
+        msg = "Length of clean names do not match length of original names"
+    )
+
+    # update the database if there is a valid internet connection
+    if (curl::has_internet()) {
+        taxadb::td_create(
+            provider = database,
+            overwrite = FALSE
+        )
+    }
+
+    # add the database to the data
+    data$db <- ifelse(is.na(data[[clean.col]]), NA, database)
+
+    # subset out taxa with special names
+    spec.names <- special_taxon_names()
+    data.spec <- dplyr::filter(
+        data,
+        (eval(parse(text = clean.col)) %in% spec.names)
+    )
+
+    # change the database column to special
+    if (nrow(data.spec) > 0) {
+        data.spec[["db"]] <- "special"
+    }
+
+    # remove the special names from the data
+    data <- dplyr::filter(data, !(row_id %in% data.spec[["row_id"]]))
```

Comment:
@haganjam R complains that there is no variable called `row_id` - I know that often you can refer to columns etc. just by their name without quotes - but probably because it gets interpreted as a string?

So can we replace this with:
```suggestion
    data <- dplyr::filter(data, !("row_id" %in% data.spec[["row_id"]]))
```
? Is it the same semantically?

Reply:
Semantically it should be the same. Does it work if you change it to a string?

Reply:
Otherwise, it might be that there is a problem with the code where row_id is actually not in the data but I haven't had that error before.